### PR TITLE
fix(s57): MBTiles metadata patch was producing duplicate rows (1.12.1-beta.1)

### PR DESCRIPTION
**Pull request by dirkwa**
## Summary
Fixes a real bug in the post-tippecanoe MBTiles metadata patcher that's been silently producing files with `type=overlay` (tippecanoe's default) instead of `type=S-57` for some users since 1.10. Bumps to **1.12.1-beta.1** for field validation before stable.

## The bug (two parts, one user-visible symptom)

A user on Discord reported that their converted `.mbtiles` files were showing tippecanoe's defaults:

```
$ sqlite3 DE_ENCs.mbtiles "SELECT name, value FROM metadata WHERE name IN ('type','name','format');"
format|pbf
name|/output/DE_ENCs.mbtiles
type|overlay
```

Investigation found two distinct bugs in the existing patcher:

1. **`INSERT OR REPLACE` was an append**. Tippecanoe creates the `metadata` table *without* a UNIQUE constraint on `name`, so `INSERT OR REPLACE` doesn't replace — it just inserts a new row. Their file had two `type` rows (tippecanoe's `overlay` plus our `S-57`) and SELECT order decided which their reader saw. Reproducible:
   ```
   sqlite> CREATE TABLE metadata (name TEXT, value TEXT);
   sqlite> INSERT INTO metadata VALUES ('type', 'overlay');
   sqlite> INSERT OR REPLACE INTO metadata VALUES ('type', 'S-57');
   sqlite> SELECT * FROM metadata;
   type|overlay
   type|S-57       ← we appended, didn't replace
   ```
2. **Failures were silent**. The patcher was wrapped in try/catch and logged only via `app.debug()`. So even when the patcher *did* fail outright (e.g. `node:sqlite` unavailable on older Node, lock contention with a concurrent reader), nothing surfaced in the conversion log or server stdout.

## The fix
Extract the patcher into a testable helper `patchS57Mbtiles` that:

- **DELETEs the rows it owns before INSERTing**. Avoids the duplicate-row failure mode entirely. Also self-heals files we already wrote with the broken patcher.
- **Probes `node:sqlite` availability up front** and reports the precise reason it isn't usable.
- **Retries once after a 500ms delay** on transient failure (rides out sqlite locks from a brief concurrent reader).
- **Verifies the write** by reading back `type` and `name` and comparing.
- **Logs every step to both** `appendLog` (visible in conversion UI) **and** `console.warn` for failures (visible in server stdout regardless of debug flag).

Best-effort contract preserved: never throws. The `.mbtiles` is still functional as a chart even if the patch fails — only metadata is affected, and now the user can see why.

## Files
- `src/utils/mbtiles-metadata.ts` (new) — `patchS57Mbtiles`, `loadSqlite`, `PatchResult`, `PatchOptions`. Pure-ish, dependency-injectable for tests.
- `test/mbtiles-metadata.test.js` (new) — 9 tests using real `node:sqlite` against synthetic MBTiles fixtures.
- `src/utils/s57-converter.ts` — replace 14-line inline patcher with a single `patchS57Mbtiles` call.
- `package.json` — bump to `1.12.1-beta.1`.

## Tested
- `npm run format:check && npm run build && npm test` — 126/126 pass (was 117). 9 new tests:
  - Happy path: tippecanoe defaults → S-57 + custom name.
  - `chartNumber=''` → falls back to `S-57 ENC`.
  - Empty metadata table → inserts cleanly.
  - Missing file → `ok=false`, `attempts=0`, clear message.
  - Open failure → `ok=false`, no throw.
  - Transient-failure retry → second attempt succeeds, retry is logged.
  - Verify-after-write structure (smoke).
  - **Regression guard**: no duplicate rows after starting from tippecanoe defaults.
  - **Self-heal**: collapses pre-existing duplicate rows from the broken patcher.
- Local `cr review --plain` — clean, no findings.
- End-to-end via the existing converter wiring; `s57-converter.ts` calls the helper exactly where the inline patcher used to live.

## Summary by CodeRabbit

* **New Features**
  * Added an S-57 MBTiles metadata patcher that standardizes chart metadata and reports progress/messages to the UI.

* **Bug Fixes**
  * Improved retry, verification and diagnostic behavior for transient update failures; failures now surface clearer warnings without throwing.

* **Tests**
  * Added comprehensive tests covering metadata updates, retries, edge cases and resilience.

* **Chores**
  * Bumped package version to 1.12.1-beta.1.

### CODE REVIEW

File: src/utils/mbtiles-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        log(message);
+        return {
+          ok: true,
+          type: gotType,
+          name: gotName,
+          attempts: attempt,
+          message
+        };
+      } finally {
+        db.close();
```
[the remark this answers is not in the document]
Comment (dirkwa) [reply src/utils/mbtiles-metadata.ts:193]: @coderabbitai Fair catch — without a transaction a verify failure after the DELETE would have left the file with NO type/name rows, strictly worse than the tippecanoe defaults we started from. Wrapped the whole DELETE + INSERT + verify in `BEGIN TRANSACTION … COMMIT`, with `ROLLBACK` on verify mismatch and a finally-block `ROLLBACK` guard for any throw mid-transaction. Pushed as 63d339a, plus a test that confirms unrelated metadata rows are preserved across a successful patch.

File: src/utils/mbtiles-metadata.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        // Verify-after-write: read back what we just put. INSERT OR REPLACE
+        // should always succeed when the prepare succeeds, but a busy WAL
+        // checkpoint or a process-level write filter could swallow it
+        // silently — and that's the bug we're trying to make visible.
+        const typeRow = db.prepare("SELECT value FROM metadata WHERE name = 'type'").get() as
+          | { value?: string }
+          | undefined;
+        const nameRow = db.prepare("SELECT value FROM metadata WHERE name = 'name'").get() as
+          | { value?: string }
+          | undefined;
```
[the remark this answers is not in the document]
Comment (dirkwa) [reply src/utils/mbtiles-metadata.ts]: @coderabbitai Already fixed in commit 655a116 (pushed before this review landed). Replaced the `as { value?: string } | undefined` casts with runtime `typeof`/`'value' in` checks since CI's TypeScript already infers the row shape. Lint is now clean.

File: src/utils/mbtiles-metadata.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+  // Two attempts: first try, then a single retry after retryMs to ride out
+  // transient sqlite locks (e.g. another process briefly holding the db).
+  // Anything beyond that is unlikely to clear up by waiting longer.
+  let lastError = '';
+  for (let attempt = 1; attempt <= 2; attempt++) {
+    if (attempt === 2) {
+      log(`Retrying MBTiles metadata patch after ${retryMs}ms…`);
+      await sleeper(retryMs);
+    }
```
[the remark this answers is not in the document]
Comment (dirkwa) [reply src/utils/mbtiles-metadata.ts]: @coderabbitai Good catch on both surfaces. Fixed in 63d339a:

- `options.onMessage` is now invoked through a wrapper that swallows callback errors. A misbehaving consumer logger can no longer break the never-throws contract.
- `await sleeper(retryMs)` is wrapped in try/catch. A throwing sleep callback (test stub or runtime weirdness) is logged and treated as a no-op delay; the retry still happens.

Two new tests pin these down (`onMessage callback that throws → still completes` and `sleep callback that throws → retry still happens`).
